TASK: Workspace aware commands events migration

mhsdesign opened the pull request:
See https://github.com/neos/neos-development-collection/pull/4708

mhsdesign committed:
WIP: Write migration for workspace rename

untested.

mhsdesign committed:
TASK: Simplify `migrateevents:fillworkspacenameincommandpayloadofeventmetadata`

As the previous approach lead to a lot of dummy workspace names we will just reduce the complexity and ONLY write dummy workspace names as they are redundant either way

mhsdesign committed:
TASK: Adjust workspace event migration to preserve original cs stream id in workspace name

mhsdesign committed:
TASK: Make `ContentStreamIdOverride` code flow more obvious

ahaeslich commented on Neos.ContentRepositoryRegistry/Classes/Command/MigrateEventsCommandController.php:
```diff
@@ -32,4 +32,20 @@ public function migratePropertiesToUnsetCommand(string $contentRepository = 'def
+     *
+     * Needed for #4708: https://github.com/neos/neos-development-collection/pull/4708
+     *
+     * Included in February 2023 - before final Neos 9.0 release
```
```suggestion
     * Included in March 2024 - before final Neos 9.0 release
```

mhsdesign committed:
TASK: Update migration docs

nezaniel reviewed (approved):
Nicely done

ahaeslich reviewed (changes requested):
Tested it locally and run into this error while replaying the contentStream projection:

```
Undefined array key "workspaceName" in /application/Packages/Neos/Neos.ContentRepository.Core/Classes/Feature/WorkspacePublication/Dto/NodeIdToPublishOrDiscard.php line 46
```

ahaeslich committed:
TASK: Remove obsolete workspaceName from NodeIdToPublishOrDiscard

mhsdesign commented:
@ahaeslich looks perfect thanks, we will merge this tomorrow once we have @bwaidelich conceptional approval.

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php:
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
+    /**
+     * @internal
+     */
+    public static function findContentStreamIdForWorkspace(ContentRepository $contentRepository, WorkspaceName $workspaceName): ContentStreamId
```
nitpick: We have the guideline to use `find*` for methods that can return `null`
```suggestion
    public static function getContentStreamIdForWorkspace(ContentRepository $contentRepository, WorkspaceName $workspaceName): ContentStreamId
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php:
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
      * @internal
      */
-    public static function useContentStreamId(?ContentStreamId $contentStreamIdToUse): void
+    public static function withContentStreamIdToUse(ContentStreamId $contentStreamIdToUse, \Closure $fn): void
```
"withContentStreamToUse" is a bit bumpy.. I can't think of something better atm.
But more importantly we should add a doc comment explaining this unusual construct even though it is just used internally

kitsunet replied:
I would actually avoid `with` as it doesn't follow our usual pattern of how `with` work (aka return a new instance with the property applied).
how about `ContentStreamIdOverride::applyContentStreamId` or even `applyContentStreamIdToClosure` to be very specific.

mhsdesign replied:
yeah sounds fine the `with` pattern is actually overloaded if you think of the `withoutAuthentication` function which also accepts a closure.

I sadly havent found yet a good prefix for these kinds of functions that set state and cleanup while another function is called.

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php:
```diff
@@ -24,13 +27,41 @@ class ContentStreamIdOverride
         self::$contentStreamIdToUse = $contentStreamIdToUse;
+        try {
+            $fn();
+        } catch (\Throwable $th) {
+            self::$contentStreamIdToUse = null;
+            throw $th;
+        }
```
```suggestion
        } finally {
            self::$contentStreamIdToUse = null;
        }
```

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -222,6 +231,88 @@ public function migratePropertiesToUnset(\Closure $outputFn)
+        $this->eventsModified = [];
+
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId)
+            . '_bak_' . date('Y_m_d_H_i_s');
```
another nitpicky suggestion:
```suggestion
            . '_bkp_' . date('Y_m_d_H_i_s');
```

mhsdesign replied:
followed the pattern of another migration, please adjust also there if here ^^ // cc @ahaeslich

bwaidelich reviewed (commented):
Looks great already, just some (partly quite nitpicky) suggestions

ahaeslich commented:
For documenation: replaying all projects now worked on the projects database.

nezaniel committed:
TASK: Apply review suggestions

- Update Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php
- Update Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ContentStreamIdOverride.php:
```diff
@@ -40,9 +40,8 @@ public static function withContentStreamIdToUse(ContentStreamId $contentStreamId
             self::$contentStreamIdToUse = null;
-            throw $th;
         }
         self::$contentStreamIdToUse = null;
```
this line can now be removed, sorry forgot that
